src/xref2/compile.ml

```diff
@@ -409,196 +373,156 @@ and functor_parameter : Env.t -> FunctorParameter.t -> FunctorParameter.t =
-    match Env.lookup_module a.id env with
-    | Some f -> f
-    | None -> raise Not_found
-  in
-  let env, expn =
-    match functor_arg.type_ with
-    | ModuleType expr -> (
-        try
-          let env, ce =
-            Expand_tools.expansion_of_module_type_expr env sg_id expr
-          in
-          let e = Lang_of.(module_expansion empty sg_id ce) in
-          (env, Some e)
-        with Expand_tools.ExpandFailure `OpaqueModule -> (env, None) )
-    | _ -> failwith "error"
-  in
-  {
-    a with
-    expr = module_type_expr env (a.id :> Paths.Identifier.Signature.t) a.expr;
-    expansion = Component.Opt.map (expansion env) expn;
-  }
+  match
+    Env.lookup_module' a.id env' |> of_option ~error:`Lookup
+    >>= fun (functor_arg, env) ->
+    get_module_type_expr functor_arg.type_ >>= fun expr ->
+    let sg_id = (a.id :> Paths.Identifier.Signature.t) in
+    match Expand_tools.expansion_of_module_type_expr env sg_id expr with
+    | Ok (env, ce) ->
+        let e = Lang_of.(module_expansion empty sg_id ce) in
+        Ok (env, Some e)
+    | Error `OpaqueModule -> Ok (env, None)
+    | Error _ ->
+        lookup_failure ~what:(`Functor_parameter a.id) `Expand;
+        Ok (env, None)
+  with
+  | Ok (env, expn) ->
+      {
+        a with
+        expr =
+          module_type_expr env (a.id :> Paths.Identifier.Signature.t) a.expr;
+        expansion = Component.Opt.map (expansion env) expn;
+      }
+  | Error e ->
+      lookup_failure ~what:(`Functor_parameter a.id) e;
+      a
 
 and module_type_expr :
     Env.t -> Paths.Identifier.Signature.t -> ModuleType.expr -> ModuleType.expr
     =
-  let exception
-    Compile_module_type_expr of Tools.signature_of_module_error option
-  in
-  fun env id expr ->
-    let open ModuleType in
-    let rec inner resolve_signatures = function
-      | Signature s ->
-          if resolve_signatures then Signature (signature env s)
-          else Signature s
-      | Path p -> Path (module_type_path env p)
-      | With (expr, subs) -> (
-          let expr = inner false expr in
-          let cexpr = Component.Of_Lang.(module_type_expr empty expr) in
-          let csubs =
-            List.map Component.Of_Lang.(module_type_substitution empty) subs
+ fun env id expr ->
+  let open ModuleType in
+  let resolve_sub ~fragment_root (sg, env, subs) csub lsub =
+    let lang_of_map = Lang_of.with_fragment_root fragment_root in
+    let env = Env.add_fragment_root sg env in
+    let sg', sub' =
+      match (csub, lsub) with
+      | Component.ModuleType.ModuleEq (frag, _), ModuleEq (unresolved, decl) ->
+          let frag' =
+            match
+              Tools.resolve_mt_module_fragment env (fragment_root, sg) frag
+            with
+            | Some cfrag ->
+                `Resolved
+                  (Lang_of.Path.resolved_module_fragment lang_of_map cfrag)
+            | None ->
+                lookup_failure ~what:(`With_module frag) `Resolve;
+                unresolved
```

Comment:
In the best-effort case, I use the first argument of `lsub`. Is this correct ?

Reply:
lsub : Odoc_model.Lang.ModuleType.substitution
csub : the equivalent Odoc_model.Component.ModuleType.substitution
the first argument is the fragment representing the object, so yes, looks correct to me.

Reply:
In fact, this bit would all be a lot cleaner if we just had lsub and we converted lsub to csub in the body of the match clause. Some cleanup for later :-)